substrate/pkg/controllers/substrate/substrate.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/awslabs/kit/substrate/pkg/apis/v1alpha1"
+	"go.uber.org/multierr"
+	"k8s.io/client-go/util/workqueue"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func NewSession() *session.Session {
+	sess := session.Must(session.NewSession(&aws.Config{STSRegionalEndpoint: endpoints.RegionalSTSEndpoint}))
+	sess.Handlers.Build.PushBack(request.MakeAddToUserAgentFreeFormHandler("kit.sh"))
+	return sess
+}
+
+func NewController(ctx context.Context) *Controller {
+	session := NewSession()
+	ec2Client := ec2.New(session)
+	iamClient := iam.New(session)
+	ssmClient := ssm.New(session)
+	autoscalingClient := autoscaling.New(session)
+
+	routeTable := &routeTable{ec2Client}
+	vpc := &vpc{ec2Client}
+	subnet := &subnet{ec2Client}
+
+	return &Controller{
+		Resources: []Resource{
+			&iamRole{iamClient},
+			&iamPolicy{iamClient},
+			&iamProfile{iamClient},
+			vpc,
+			&elasticIP{ec2Client},
+			&internetGateway{ec2Client, vpc},
+			subnet,
+			&securityGroup{ec2Client},
+			&natGateway{ec2Client},
+			routeTable,
+			&routeTableAssociation{ec2Client, routeTable},
+			&launchTemplate{ec2Client, ssmClient},
+			&autoScalingGroup{ec2Client, autoscalingClient, subnet},
+		},
+	}
+}
+
+type Controller struct {
+	sync.RWMutex
+	Resources []Resource
+}
+
+type Resource interface {
+	Create(context.Context, *v1alpha1.Substrate) (reconcile.Result, error)
+	Delete(context.Context, *v1alpha1.Substrate) (reconcile.Result, error)
+}
+
+func (c *Controller) Reconcile(ctx context.Context, substrate *v1alpha1.Substrate) error {
+	start := time.Now()
+	logging.FromContext(ctx).Infof("Creating resources for %s", substrate.Name)
+	if err := c.apply(ctx, substrate); err != nil {
+		return err
+	}
+	logging.FromContext(ctx).Infof("Successfully created after %s", time.Since(start))
+	return nil
+}
+
+func (c *Controller) Finalize(ctx context.Context, substrate *v1alpha1.Substrate) error {
+	start := time.Now()
+	logging.FromContext(ctx).Infof("Deleting resources for %s", substrate.Name)
+	if err := c.apply(ctx, substrate); err != nil {
+		return err
+	}
+	logging.FromContext(ctx).Infof("Successfully deleted after %s", time.Since(start))
+	return nil
+}
+
+func (c *Controller) apply(ctx context.Context, substrate *v1alpha1.Substrate) error {
+	ctx, cancel := context.WithCancel(ctx)
+	var errs = make([]error, len(c.Resources))
+	workqueue.ParallelizeUntil(ctx, len(c.Resources), len(c.Resources), func(i int) {
+		for {
+			c.RLock()
+			mutable := substrate.DeepCopy()
+			c.RUnlock()
+			f := c.Resources[i].Create
+			if substrate.DeletionTimestamp != nil {
+				f = c.Resources[i].Delete
+			}
+			result, err := f(ctx, substrate)
+			if err != nil {
+				errs[i] = err
+				cancel()
+				return
+			}
+			c.Lock()
+			mutable.DeepCopyInto(substrate)
+			c.Unlock()
+			if !result.Requeue || result.RequeueAfter == 0 {
```

Comment:
If `RequeueAfter` is not set does it get retried immediately or is there a delay?